File: rust/crates/simcore/src/expr.rs
//! The bounded kinetics expression VM — the Rust port of `simcore.expr`
//! (Phase 9, Step 4a / decision A / decision D).
//!
//! This mirrors the **one deliberate, one-time extension** of the frozen Python
//! `simcore` core (the `expr.py` added in Phase-9 Step 2). Like the integrator, it
//! is a single frozen engine primitive: a plain-data AST + a pure evaluator +
//! [`DeclarativeFlow`] (rate-expr × stoichiometry, balanced by construction). It is
//! purely additive — no existing Rust `simcore` module changes — so every frozen
//! golden stays byte-identical.
//!
//! Why it lives in the core (not the boundary): an authored flow's rate expression is
//! evaluated **per step, inside the integrator** (once per Euler step, per stage under
//! RK4), so the evaluator must be deterministic like every other flow. *Parsing* a
//! scenario file's text into this AST is a one-time boundary act and stays in the
//! `authoring` crate (decision A). Only AST→`f64` evaluation is core.
//!
//! The grammar is **bounded, closed, and deterministic** (decision D) — a fixed,
//! finite set of primitives, no user functions / recursion / loops / I/O. Step 2
//! shipped the unambiguous arithmetic core, and Step 4a ports exactly that, no more:
//!
//!   * literals ([`Expr::Const`]);
//!   * reads: a stock amount by id ([`Expr::StockRef`]), a param by name
//!     ([`Expr::ParamRef`]), a forcing by name ([`Expr::ForcingRef`], resolved
//!     through [`Environment::get`] — #16), and the integer step `n` ([`Expr::StepN`]);
//!   * binary `+ - *` ([`Expr::BinOp`]) and unary `-` ([`Expr::Neg`]).
//!
//! **Deferred (do NOT complete the grammar).** Division and the closed function set
//! (`exp ln pow sqrt abs min max clamp monod` + bounded conditionals) are deliberately
//! *not* here — each carries a cross-port semantic choice (`x/0` is Python-raise vs
//! Rust-`inf`; `monod`/`clamp`/`ifpos` definitions) that a real frozen flow must force.
//! The [`BinaryOp`] enum literally cannot represent `/`, so an unimplemented op can
//! never silently evaluate — the analogue of the Python parser/evaluator's explicit
//! rejection (the "helpful dev adds all the ops" trap, structurally foreclosed).
//!
//! **No `dt` token, by construction.** The rate expression is the *instantaneous* rate
//! (`dt`-independent); [`DeclarativeFlow`] supplies the single `× dt` that makes the
//! per-step increment. So RK4-order-safety is structural for every authored flow. `n`
//! stays readable (`dt`-independent, safe).
//!
//! The AST is plain data (an `Expr` tree of `f64`/`String`/nested nodes), so this is a
//! mechanical mirror of the Python module and cross-port parity is tolerance-gated
//! exactly like every other transcendental (Phase-7 3-tier contract): a
//! transcendental-free authored flow is Tier-1 bit-exact.

use std::collections::BTreeMap;

use crate::environment::Environment;
use crate::error::SimError;
use crate::flow::{Flow, FlowResult, Leg};
use crate::ids::{FlowId, StockId};
use crate::state::State;

/// The closed binary-operator set Step 2 ships (unambiguous IEEE arithmetic).
///
/// An enum, not a `char`/`String`: `/` and the function set are *deferred*, and an
/// enum with only these three variants makes an unsupported op **unrepresentable** —
/// the Rust type system subsumes the Python evaluator's `raise ValueError` fallback,
/// the same "the check moved to compile time" pattern as the RNG `TypeError` guards.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
}

impl BinaryOp {
    /// The canonical symbol (`"+"` / `"-"` / `"*"`) — the parse-parity spelling.
    pub fn symbol(self) -> &'static str {
        match self {
            BinaryOp::Add => "+",
            BinaryOp::Sub => "-",
            BinaryOp::Mul => "*",
        }
    }
}

/// The bounded-grammar rate-expression AST (plain data — immutable, `PartialEq`).
///
/// Every node is a value type, so the whole tree is comparable and ports mechanically.
/// `Neg`/`BinOp` box their operands because the type is recursive.
#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    /// A numeric literal (the parser has already coerced it to `f64`).
    Const(f64),
    /// A read of a stock's current amount, by id, from the evaluation snapshot —
    /// read **directly** (`snapshot.stocks[id].amount`), the donor-controlled idiom
    /// (`SelfDischarge` reads `battery` this way), *not* via `env`.
    StockRef(StockId),
    /// A read of a flow param by name (from the flow's own param map).
    ParamRef(String),
    /// A read of a forcing var by name, resolved through [`Environment::get`] (#16) —
    /// indistinguishable at evaluation time from a coupled sibling's shared stock.
    ForcingRef(String),
    /// A read of the integer step count `n` (as an `f64`); `dt`-independent.
    StepN,
    /// Unary negation `- operand`.
    Neg(Box<Expr>),
    /// A binary arithmetic op `left <op> right`. `left` is evaluated before `right`
    /// and then combined — a fixed op-order the Python port mirrors, so the result is
    /// bit-identical within a build.
    BinOp {
        op: BinaryOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
}

/// Evaluate `node` to an `f64` against a snapshot/env/param context.
///
/// Pure and deterministic in its inputs. Reference resolution mirrors the frozen flows
/// and the Python `eval_expr`: [`Expr::StockRef`] reads the snapshot directly (#16
/// donor read), [`Expr::ForcingRef`] goes through [`Environment::get`] (#16
/// forcing/shared read), [`Expr::ParamRef`] reads the flow's param map, [`Expr::StepN`]
/// reads `snapshot.n`. A `StockRef`/`ParamRef` at a missing id returns
/// [`SimError::Reference`] (Python raises `KeyError`); referential integrity is
/// validated at *build* time by the interpreter, so these are belt-and-suspenders.
///
/// `dt` is intentionally absent from the signature: the rate grammar has no `dt` token,
/// so a rate expression cannot depend on `dt`.
pub fn eval_expr(
    node: &Expr,
    snapshot: &State,
    env: &dyn Environment,
    params: &BTreeMap<String, f64>,
) -> Result<f64, SimError> {
    match node {
        Expr::Const(value) => Ok(*value),
        Expr::StockRef(stock) => snapshot
            .stocks
            .get(stock)
            .map(|s| s.amount)
            .ok_or_else(|| {
                SimError::Reference(format!("rate expression reads unknown stock {stock:?}"))
            }),
        Expr::ParamRef(name) => params.get(name).copied().ok_or_else(|| {
            SimError::Reference(format!("rate expression reads unknown param {name:?}"))
        }),
        Expr::ForcingRef(name) => env.get(name),
        Expr::StepN => Ok(snapshot.n as f64),
        Expr::Neg(operand) => Ok(-eval_expr(operand, snapshot, env, params)?),
        Expr::BinOp { op, left, right } => {
            // left before right, then combine — the fixed op-order Python mirrors.
            let l = eval_expr(left, snapshot, env, params)?;
            let r = eval_expr(right, snapshot, env, params)?;
            Ok(match op {
                BinaryOp::Add => l + r,
                BinaryOp::Sub => l - r,
                BinaryOp::Mul => l * r,
            })
        }
    }
}

/// An authored [`Flow`]: an instantaneous *rate* expression × a fixed stoichiometry —
/// the Rust port of Python's `DeclarativeFlow`.
///
/// **Balanced by construction (decision C).** The flow emits one leg per
/// `(stock, coeff)` pair, all sharing the single scalar `increment = rate·dt`, so per
/// conserved quantity `Σ legs = rate·dt · Σ(coeff · composition)` — which is `0` for
/// *any* rate value **iff** the stoichiometry's coefficient vector balances. The
/// author picks only the scalar rate and the (integer/rational) coefficients; they
/// cannot vary the per-leg magnitude independently. The interpreter validates the
/// coefficient vector against the stock compositions at build time; the every-step
/// conservation gate is then a redundant backstop.
///
/// **Increment-form (RK4-order-safe).** `rate` is the `dt`-independent instantaneous
/// rate; [`Flow::evaluate`] forms `increment = rate · dt` and each leg is
/// `coeff · increment` — the standard flux→`× dt` split every frozen flow uses, so
/// RK4's ⅙-combine reproduces classical RK4 exactly. The rate has no `dt` token
/// (grammar-enforced), so `dt`-linearity is structural.
///
/// `params` is stored as a sorted `(name, value)` list (the frozen-flow idiom) and
/// exposed to the evaluator as a `BTreeMap`; `stoichiometry` is `(stock, coeff)` in
/// author order (leg order does not affect the trajectory — each leg lands on a
/// distinct stock and the integrator sums per stock).
#[derive(Debug, Clone, PartialEq)]
pub struct DeclarativeFlow {
    id: FlowId,
    priority: i64,
    rate: Expr,
    stoichiometry: Vec<(StockId, f64)>,
    params: Vec<(String, f64)>,
}

impl DeclarativeFlow {
    /// Construct a `DeclarativeFlow`. `params` is stored sorted by name (the
    /// frozen-flow idiom + a stable canonical form); `stoichiometry` keeps author
    /// order (leg order is inert).
    pub fn new(
        id: FlowId,
        priority: i64,
        rate: Expr,
        stoichiometry: Vec<(StockId, f64)>,
        params: Vec<(String, f64)>,
    ) -> DeclarativeFlow {
        let mut params = params;
        params.sort_by(|a, b| a.0.cmp(&b.0));
        DeclarativeFlow {
            id,
            priority,
            rate,
            stoichiometry,
            params,
        }
    }

    fn param_map(&self) -> BTreeMap<String, f64> {
        self.params.iter().cloned().collect()
    }
}

impl Flow for DeclarativeFlow {
    fn id(&self) -> &str {
        &self.id
    }

    fn priority(&self) -> i64 {
        self.priority
    }

    fn evaluate(
        &self,
        snapshot: &State,
        env: &dyn Environment,
        dt: f64,
    ) -> Result<FlowResult, SimError> {
        let rate = eval_expr(&self.rate, snapshot, env, &self.param_map())?;
        let increment = rate * dt;
        let mut legs = Vec::with_capacity(self.stoichiometry.len());
        for (stock, coeff) in &self.stoichiometry {
            legs.push(Leg::new(stock.clone(), coeff * increment)?);
        }
        FlowResult::new(legs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::environment::{constant, SourceResolver};
    use crate::flow::assert_flow_balanced_default;
    use crate::quantities::{Quantity, StockKind};
    use crate::state::Stock;

    const BATTERY: &str = "power.battery";
    const WASTE_HEAT: &str = "boundary.waste_heat";

    fn energy_stock(id: &str, amount: f64, kind: StockKind) -> Stock {
        Stock::new(
            id.to_string(),
            "power".to_string(),
            Quantity::Energy,
            "J".to_string(),
            amount,
            kind,
            0.0,
            false,
            BTreeMap::new(),
        )
        .unwrap()
    }

    fn state(battery: f64, n: u64) -> State {
        let mut stocks = BTreeMap::new();
        stocks.insert(BATTERY.to_string(), energy_stock(BATTERY, battery, StockKind::Pool));
        stocks.insert(
            WASTE_HEAT.to_string(),
            energy_stock(WASTE_HEAT, 0.0, StockKind::Boundary),
        );
        State::new(n, stocks, 0, BTreeMap::new()).unwrap()
    }

    fn eval(node: &Expr, battery: f64, n: u64, params: &[(&str, f64)]) -> f64 {
        let st = state(battery, n);
        let resolver = SourceResolver::empty();
        let env = resolver.bind(&st, 0.0);
        let pmap: BTreeMap<String, f64> =
            params.iter().map(|(k, v)| (k.to_string(), *v)).collect();
        eval_expr(node, &st, &env, &pmap).unwrap()
    }

    // --- leaf reads --------------------------------------------------------
    #[test]
    fn const_evaluates_to_its_value() {
        assert_eq!(eval(&Expr::Const(2.5), 5.0, 3, &[]), 2.5);
    }

    #[test]
    fn stock_ref_reads_the_snapshot_amount() {
        assert_eq!(eval(&Expr::StockRef(BATTERY.to_string()), 7.25, 3, &[]), 7.25);
    }

    #[test]
    fn param_ref_reads_the_param_map() {
        assert_eq!(
            eval(&Expr::ParamRef("k".to_string()), 5.0, 3, &[("k", 1.0e-8)]),
            1.0e-8
        );
    }

    #[test]
    fn forcing_ref_reads_through_env() {
        // A forcing var resolves through env.get (#16); build a resolver that has it.
        let st = state(5.0, 3);
        let mut forcings = std::collections::HashMap::new();
        forcings.insert("load_power".to_string(), constant(42.0).unwrap());
        let resolver = SourceResolver::new(forcings, std::collections::HashMap::new()).unwrap();
        let env = resolver.bind(&st, 0.0);
        let pmap = BTreeMap::new();
        let value = eval_expr(&Expr::ForcingRef("load_power".to_string()), &st, &env, &pmap);
        assert_eq!(value.unwrap(), 42.0);
    }

    #[test]
    fn step_n_reads_n_as_float() {
        assert_eq!(eval(&Expr::StepN, 5.0, 9, &[]), 9.0);
    }

    #[test]
    fn missing_param_is_a_reference_error() {
        // Belt-and-suspenders (the interpreter validates refs at build time): the raw
        // VM surfaces a missing param as SimError::Reference, not a wrong answer.
        let st = state(5.0, 3);
        let resolver = SourceResolver::empty();
        let env = resolver.bind(&st, 0.0);
        let pmap = BTreeMap::new();
        let err = eval_expr(&Expr::ParamRef("absent".to_string()), &st, &env, &pmap);
        assert!(matches!(err, Err(SimError::Reference(_))));
    }

    // --- operators, op-for-op ---------------------------------------------
    fn binop(op: BinaryOp, left: Expr, right: Expr) -> Expr {
        Expr::BinOp {
            op,
            left: Box::new(left),
            right: Box::new(right),
        }
    }

    #[test]
    fn addition() {
        assert_eq!(
            eval(&binop(BinaryOp::Add, Expr::Const(0.1), Expr::Const(0.2)), 5.0, 3, &[]),
            0.1 + 0.2
        );
    }

    #[test]
    fn subtraction() {
        assert_eq!(
            eval(&binop(BinaryOp::Sub, Expr::Const(1.0), Expr::Const(0.3)), 5.0, 3, &[]),
            1.0 - 0.3
        );
    }

    #[test]
    fn multiplication() {
        let expr = binop(
            BinaryOp::Mul,
            Expr::Const(1.0e-8),
            Expr::StockRef(BATTERY.to_string()),
        );
        assert_eq!(eval(&expr, 1.0e7, 3, &[]), 1.0e-8 * 1.0e7);
    }

    #[test]
    fn unary_negation() {
        assert_eq!(
            eval(&Expr::Neg(Box::new(Expr::StockRef(BATTERY.to_string()))), 3.0, 3, &[]),
            -3.0
        );
    }

    #[test]
    fn negation_is_exact_for_zero() {
        // -x sign-of-zero: what makes DeclarativeFlow's leg formula bit-identical to a
        // frozen `-leak`; assert the VM's Neg produces -0.0.
        let v = eval(&Expr::Neg(Box::new(Expr::Const(0.0))), 5.0, 3, &[]);
        assert_eq!(v.signum(), -1.0);
        assert!(v == 0.0 && v.is_sign_negative());
    }

    #[test]
    fn left_associativity_of_subtraction() {
        // (10 - 3) - 2, not 10 - (3 - 2): left combined before right.
        let expr = binop(
            BinaryOp::Sub,
            binop(BinaryOp::Sub, Expr::Const(10.0), Expr::Const(3.0)),
            Expr::Const(2.0),
        );
        assert_eq!(eval(&expr, 5.0, 3, &[]), (10.0 - 3.0) - 2.0);
    }

    #[test]
    fn nested_rate_expression_op_for_op() {
        // k * (battery - n): a demand-shaped rate, in the fixed op-order.
        let expr = binop(
            BinaryOp::Mul,
            Expr::ParamRef("k".to_string()),
            binop(BinaryOp::Sub, Expr::StockRef(BATTERY.to_string()), Expr::StepN),
        );
        assert_eq!(eval(&expr, 8.0, 2, &[("k", 0.5)]), 0.5 * (8.0 - 2.0));
    }

    // --- DeclarativeFlow ---------------------------------------------------
    fn self_discharge_flow(k: f64) -> DeclarativeFlow {
        // rate k·battery, −1/+1 ENERGY split (the SelfDischarge re-expression).
        DeclarativeFlow::new(
            "power.self_discharge".to_string(),
            0,
            binop(
                BinaryOp::Mul,
                Expr::ParamRef("k".to_string()),
                Expr::StockRef(BATTERY.to_string()),
            ),
            vec![(BATTERY.to_string(), -1.0), (WASTE_HEAT.to_string(), 1.0)],
            vec![("k".to_string(), k)],
        )
    }

    fn evaluate_flow(flow: &DeclarativeFlow, battery: f64, dt: f64) -> FlowResult {
        let st = state(battery, 0);
        let resolver = SourceResolver::empty();
        let env = resolver.bind(&st, dt);
        flow.evaluate(&st, &env, dt).unwrap()
    }

    #[test]
    fn declarative_flow_legs_and_increment_form() {
        let flow = self_discharge_flow(1.0e-8);
        let result = evaluate_flow(&flow, 1.0e7, 3600.0);
        let leak = (1.0e-8 * 1.0e7) * 3600.0;
        // Legs in author (stoichiometry) order; each on a distinct stock.
        let stocks: Vec<&str> = result.legs.iter().map(|l| l.stock.as_str()).collect();
        assert_eq!(stocks, vec![BATTERY, WASTE_HEAT]);
        let battery_leg = result.legs.iter().find(|l| l.stock == BATTERY).unwrap();
        let heat_leg = result.legs.iter().find(|l| l.stock == WASTE_HEAT).unwrap();
        assert_eq!(battery_leg.amount, -leak);
        assert_eq!(heat_leg.amount, leak);
    }

    #[test]
    fn declarative_flow_is_energy_balanced() {
        let flow = self_discharge_flow(1.0e-8);
        let st = state(1.0e7, 0);
        let result = evaluate_flow(&flow, 1.0e7, 3600.0);
        // Errors if unbalanced; the −1/+1 split balances exactly.
        assert_flow_balanced_default(&result, &st.stocks).unwrap();
    }

    #[test]
    fn declarative_flow_is_dt_linear() {
        let flow = self_discharge_flow(1.0e-8);
        let small = evaluate_flow(&flow, 1.0e7, 1.0);
        let big = evaluate_flow(&flow, 1.0e7, 10.0);
        let small_b = small.legs.iter().find(|l| l.stock == BATTERY).unwrap().amount;
        let big_b = big.legs.iter().find(|l| l.stock == BATTERY).unwrap().amount;
        // The rate is dt-independent, so the increment scales exactly linearly in dt.
        assert_eq!(big_b, small_b * 10.0);
    }

    #[test]
    fn declarative_flow_zero_battery_is_no_op() {
        let flow = self_discharge_flow(1.0e-8);
        let result = evaluate_flow(&flow, 0.0, 3600.0);
        for leg in &result.legs {
            assert_eq!(leg.amount, 0.0);
        }
    }
}

File: src/simcore/expr.py
"""The bounded kinetics expression VM: a plain-data AST + evaluator + DeclarativeFlow.

**This module is the one deliberate, one-time extension of the frozen ``simcore``
core since the biosphere/station freeze** (Phase 9, decision A / decision D). It is
purely *additive* — no existing ``simcore`` file changes — so every frozen golden
stays byte-identical; the "zero core change" streak is broken **once, on purpose**,
to add a single frozen engine primitive (like adding a new integrator), never
per-scenario code.

Why it lives in the core (not the boundary): an authored flow's rate expression is
evaluated **per step, inside the integrator** (once per Euler step, per stage under
RK4), so the evaluator must be pure stdlib and deterministic like every other flow.
*Parsing* a scenario file's text into this AST is a one-time boundary act and stays
in ``src/authoring`` (decision A). Only AST→``float`` evaluation is core.

The grammar is **bounded, closed, and deterministic** (decision D) — a fixed, finite
set of primitives, no user functions / recursion / loops / I/O. Step 2 ships the
unambiguous arithmetic core exercised by the re-expression anchor (``SelfDischarge``):

  * literals (``Const``);
  * reads: a stock amount by id (``StockRef``), a param by name (``ParamRef``), a
    forcing by name (``ForcingRef``, resolved through ``env.get`` — #16), and the
    integer step ``n`` (``StepN``);
  * binary ``+ - *`` (``BinOp``) and unary ``-`` (``Neg``);
  * saturating kinetics ``monod`` (``Monod``) — the post-roadmap Tier-2 unfreeze; see
    that node's docstring for the frozen flow that forced its definition.

**Deliberately deferred** until a real frozen flow forces the semantic definition
(the ``_DemoSchema``/``_CanopySchema`` "bespoke until a second instance justifies it"
discipline, applied to the grammar):

  * ``/`` — division is IEEE-unambiguous *except* at ``x/0`` (Python raises
    ``ZeroDivisionError``; Rust f64 yields ``inf``). ``monod`` (the first divider)
    has now landed **without** lifting this: it guards its own denominator, so it
    resolves ``x/0`` *internally* and never exposes the raw form. Bare ``/`` would
    re-introduce exactly the hazard this defers, and no frozen flow forces it;
  * the rest of the closed **function set** — ``exp ln pow sqrt abs min max clamp`` +
    bounded conditionals. ``clamp``/``ifpos`` still carry a real definitional choice
    (inclusive bounds? ``>0`` vs ``≥0``?), and a faithful transcendental anchor
    additionally needs a **named-constant** surface (Stefan-Boltzmann's σ is a CODATA
    *module constant*, not a param — see ``domains.thermal``), unresolved here. Kept
    explicit so the Rust port does not assume the grammar is complete.

**No ``dt`` token, by construction.** Exposing ``dt`` to the rate grammar would let an
author write ``coeff·dt·f(dt)`` — non-linear in ``dt`` — silently forfeiting RK4 order
(the very thing the increment-form contract in ``simcore.flow`` protects). The rate
expression is the **instantaneous rate** (``dt``-independent); :class:`DeclarativeFlow`
supplies the single ``× dt`` that makes the per-step increment. So RK4-order-safety is
*structural* for every authored flow, not a documented hope. ``n`` stays readable
(``dt``-independent, safe).

Pure stdlib only. The AST is plain data (frozen dataclasses of ``float``/``str``/nested
nodes), so the Step-4 Rust port is a mechanical mirror and cross-port parity is
tolerance-gated exactly like every other transcendental (Phase-7 3-tier contract): a
transcendental-free authored flow is Tier-1 bit-exact.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass

from simcore.environment import Environment
from simcore.flow import FlowResult, Leg
from simcore.ids import FlowId, StockId
from simcore.state import State

# --- the AST (plain data) ----------------------------------------------------
# Every node is a frozen dataclass of primitives / nested nodes, so the whole tree
# is immutable + hashable + equatable and ports mechanically to Rust. ``Expr`` is the
# closed node union (Step 2's arithmetic core); the evaluator dispatches on it.


@dataclass(frozen=True)
class Const:
    """A numeric literal (a ``float`` — the parser has already coerced it)."""

    value: float


@dataclass(frozen=True)
class StockRef:
    """A read of a stock's current amount, by id, from the evaluation snapshot.

    Reads the snapshot **directly** (``snapshot.stocks[id].amount``), the
    donor-controlled idiom (``SelfDischarge`` reads ``battery`` this way), *not* via
    ``env`` — a stock read and a forcing read are distinct grammar forms.
    """

    stock: StockId


@dataclass(frozen=True)
class ParamRef:
    """A read of a flow param by name (from the flow's own param map)."""

    name: str


@dataclass(frozen=True)
class ForcingRef:
    """A read of a forcing var by name, resolved through ``env.get`` (#16).

    Indistinguishable at evaluation time from a coupled sibling's shared stock —
    exactly ``env``'s contract; a forced authored flow reads its rate this way.
    """

    name: str


@dataclass(frozen=True)
class StepN:
    """A read of the integer step count ``n`` (as a ``float``); ``dt``-independent."""


@dataclass(frozen=True)
class Neg:
    """Unary negation ``- operand``."""

    operand: Expr


@dataclass(frozen=True)
class BinOp:
    """A binary arithmetic op ``left <op> right`` with ``op`` ∈ ``{"+", "-", "*"}``.

    ``left`` is evaluated before ``right`` and then combined — a fixed op-order the
    Rust port mirrors, so the result is bit-identical within a build.
    """

    op: str
    left: Expr
    right: Expr


@dataclass(frozen=True)
class Monod:
    """Saturating kinetics ``substrate / (substrate + half_saturation)``.

    The **one shape ``+ − ×`` cannot approximate**, and the most common functional form
    in the science this project models — Michaelis-Menten, Monod and Holling type II all
    share this algebra. Added by the post-roadmap Tier-2 grammar unfreeze
    (``docs/plans/post-roadmap-grammar-monod.md``).

    **A frozen flow forced the definition** (the grammar's own precondition): this is
    the kernel of ``domains.biosphere.chamber.oxygen_limitation_factor`` — frozen since
    Phase-2 Step 7, cited to Davidson et al. 2012, and used by three frozen flows. Both
    of the choices the grammar had to make were therefore *already made, and cited*:

    * **2-arg, dimensionless** (not a 3-arg ``Vmax·S/(S+K)``). The frozen ``f_O2`` is a
      dimensionless factor applied as ``daily · f_O2 · dt``, so ``Vmax`` arrives through
      the already-frozen ``*``. Arg order is ``monod(substrate, half_saturation)``,
      matching the frozen signature and Michaelis-Menten convention — a **frozen
      semantic choice**, pinned by a parse vector.
    * **``denom <= 0`` → ``0.0``**, mirroring the frozen line-for-line ("no O₂ ⇒ no
      respiration"). This makes the node **total**: for all finite inputs it returns a
      finite float — never NaN, never ±inf, never raising. So ``0/0`` cannot reach a
      hex-float golden, and the Python-raise-vs-Rust-``inf`` split never arises because
      no raw ``x/0`` is reachable.

    **Only the kernel is mirrored, not the frozen function's argument preparation.** The
    frozen ``max(0.0, o2_mol)/air_mol`` and its ``ValueError`` guards are *arg prep* for
    a depleting physical pool; for an authored rate that layer **is the sub-expressions
    the author composed**. Baking a silent ``max(0, ·)`` in here would change
    ``monod(stock("a") - stock("b"), k)`` invisibly — the silent-failure class this
    platform exists to avoid. Off the natural domain (``S ≥ 0``, ``K > 0``, where the
    result is the textbook ``[0, 1)`` and bit-identical to frozen ``f_O2``) the node
    stays total and yields conservation-closed nonsense the author owns, by design.

    **Tier-1 bit-exact, not Tier-2.** Division is an IEEE-754 *basic* operation —
    correctly-rounded and deterministic cross-port, exactly like ``+ − *``. The
    "a transcendental moves an authored flow Tier-1 → Tier-2" rule covers *libm* calls
    (``RadiatorReject``'s ``T⁴`` via ``pow``); this is not in that class.

    **RK4-order-safe.** ``S/(S+K)`` is C∞ on the natural domain, so RK4's convergence
    order survives — which the obvious cheap alternative ``min(k·S, Vmax)`` would
    destroy (its kink is non-differentiable). The ``denom <= 0`` branch is a derivative
    discontinuity only at the pathological boundary, which ``K > 0`` never reaches.
    """

    substrate: Expr
    half_saturation: Expr


Expr = Const | StockRef | ParamRef | ForcingRef | StepN | Neg | BinOp | Monod

# The op set Step 2 ships (unambiguous IEEE arithmetic). Division and the function
# set are deferred (see the module docstring); the parser and evaluator both reject
# anything outside this set, so an unimplemented op can never silently evaluate.
_BINARY_OPS: frozenset[str] = frozenset({"+", "-", "*"})


def eval_expr(
    node: Expr,
    snapshot: State,
    env: Environment,
    params: Mapping[str, float],
) -> float:
    """Evaluate ``node`` to a ``float`` against a snapshot/env/param context.

    Pure and deterministic in its inputs. Reference resolution mirrors the frozen
    flows: ``StockRef`` reads the snapshot directly (#16 donor read), ``ForcingRef``
    goes through ``env.get`` (#16 forcing/shared read), ``ParamRef`` reads the flow's
    param map, ``StepN`` reads ``snapshot.n``. A ``StockRef`` at a missing id raises
    ``KeyError`` and a ``ParamRef`` at a missing name raises ``KeyError`` — referential
    integrity is validated at *build* time by the interpreter (an ``AuthoringError``),
    so these are belt-and-suspenders, never the primary guard.

    ``dt`` is intentionally absent from the signature: the rate grammar has no ``dt``
    token (see the module docstring), so a rate expression cannot depend on ``dt``.
    """
    if isinstance(node, Const):
        return node.value
    if isinstance(node, StockRef):
        return snapshot.stocks[node.stock].amount
    if isinstance(node, ParamRef):
        return params[node.name]
    if isinstance(node, ForcingRef):
        return env.get(node.name)
    if isinstance(node, StepN):
        return float(snapshot.n)
    if isinstance(node, Neg):
        return -eval_expr(node.operand, snapshot, env, params)
    if isinstance(node, BinOp):
        left = eval_expr(node.left, snapshot, env, params)
        right = eval_expr(node.right, snapshot, env, params)
        if node.op == "+":
            return left + right
        if node.op == "-":
            return left - right
        if node.op == "*":
            return left * right
        # Unreachable for a well-formed AST (the parser only emits _BINARY_OPS and the
        # interpreter re-validates), but kept explicit so a malformed op is a loud
        # error rather than a silent wrong answer.
        raise ValueError(f"unsupported binary op {node.op!r}")
    if isinstance(node, Monod):
        # substrate before half_saturation — the same fixed left-to-right order BinOp
        # uses, mirrored by the Rust port.
        substrate = eval_expr(node.substrate, snapshot, env, params)
        half_saturation = eval_expr(node.half_saturation, snapshot, env, params)
        denom = substrate + half_saturation
        if denom <= 0.0:
            # The frozen f_O2's own choice, verbatim ("no O₂ ⇒ no respiration"): the
            # degenerate 0/0 returns 0 rather than NaN, which is what makes this node
            # total. Also catches a negative denominator (an author's negative K), where
            # S/denom would otherwise flip sign.
            return 0.0
        return substrate / denom
    raise TypeError(f"not an Expr node: {node!r}")  # pragma: no cover - exhaustive


@dataclass(frozen=True)
class DeclarativeFlow:
    """An authored ``Flow``: an instantaneous *rate* expression × a fixed stoichiometry.

    **Balanced by construction (decision C).** The flow emits one leg per
    ``(stock, coeff)`` pair, all sharing the single scalar ``increment = rate·dt``, so
    per conserved quantity ``Σ legs = rate·dt · Σ(coeff · composition)`` — which is
    ``0`` for *any* rate value **iff** the stoichiometry's coefficient vector balances.
    The author picks only the scalar rate and the (integer/rational) coefficients; they
    cannot vary the per-leg magnitude independently, so a *balanced* coefficient vector
    stays balanced at every step regardless of state. The interpreter validates that
    coefficient vector against the stock compositions at build time (relative tolerance,
    mirroring ``assert_flow_balanced``); the every-step conservation gate is then a
    redundant backstop. (For integer coefficients like ``−1/+1`` the balance is exact;
    for fractional split coefficients ``f/(1−f)`` it holds to floating tolerance, like
    the frozen ``charge_split`` — so "structural" is exact for integer coeffs, and
    tolerance-backed for fractional ones.)

    **Increment-form (RK4-order-safe).** ``rate`` is the ``dt``-independent
    instantaneous rate; ``evaluate`` forms ``increment = rate · dt`` and each leg is
    ``coeff · increment``. This is the standard ``flux → × dt`` split every frozen flow
    uses, so RK4's ⅙-combine reproduces classical RK4 exactly. ``rate`` has no ``dt``
    token (grammar-enforced), so ``dt``-linearity is structural.

    ``params`` is stored as a sorted ``(name, value)`` tuple (immutable + hashable, the
    frozen-flow idiom) and exposed to the evaluator as a mapping; ``stoichiometry`` is a
    ``(stock, coeff)`` tuple in author order (leg order does not affect the trajectory —
    each leg lands on a distinct stock and the integrator sums per stock).
    """

    id: FlowId
    priority: int
    rate: Expr
    stoichiometry: tuple[tuple[StockId, float], ...]
    params: tuple[tuple[str, float], ...]

    def _param_map(self) -> dict[str, float]:
        return {name: value for name, value in self.params}

    def evaluate(self, snapshot: State, env: Environment, dt: float) -> FlowResult:
        rate = eval_expr(self.rate, snapshot, env, self._param_map())
        increment = rate * dt
        return FlowResult(
            legs=tuple(
                Leg(stock, coeff * increment) for stock, coeff in self.stoichiometry
            )
        )
